R/verify.R

```diff
@@ -0,0 +1,85 @@
+#' Check Code Reproducibility
+#'
+#' Checks whether code in `teal_data` reproduces the stored objects.
+#'
+#' @return If the objects returned by the `@code` slot in the input `teal_data` object exactly match
+#' the objects in `teal_data@env`, the function updates the `@verified` slot to `TRUE` in the `teal_data` object.
+#' Once it's verified, the slot will always be set to `TRUE`. If the `@code` fails to recreate objects in
+#' `teal_data@env`, an error is thrown.
+#'
+#' @param x `teal_data` object
+#' @examples
+#' tdata1 <- teal_data()
+#' tdata1 <- within(tdata1, {
+#'   a <- 1
+#'   b <- a^5
+#'   c <- list(x = 2)
+#' })
+#' verify(tdata1)
+#'
+#' tdata2 <- teal_data(x1 = iris, code = "x1 = iris")
+#' verify(tdata2)
+#' verify(tdata2)@verified
+#' tdata2@verified
+#'
+#' \dontrun{
+#' tdata3 <- teal_data()
+#' tdata3 <- within(tdata3, {
+#'   stop("error")
+#' })
+#' verify(tdata3)
+#' }
+#'
+#' @name verify
+#' @rdname verify
+#' @aliases verify,teal_data-method
+#' @aliases verify,qenv.error-method
+#'
+#' @export
+setGeneric("verify", function(x) standardGeneric("verify"))
+setMethod("verify", "teal_data", definition = function(x) {
+  if (x@verified) {
+    return(x)
+  }
+  new_teal_data <- eval_code(teal_data(), get_code(x))
+
+  if (inherits(new_teal_data, "qenv.error")) {
+    stop(conditionMessage(new_teal_data), call. = FALSE)
+  }
+
+  reproduced <- isTRUE(all.equal(x@env, new_teal_data@env))
+  if (reproduced) {
+    x@verified <- TRUE
+    methods::validObject(x)
+    x
+  } else {
+    names_diff <- setdiff(names(x@env), names(new_teal_data@env))
+
+    objects_diff <- vapply(
+      intersect(names(x@env), names(new_teal_data@env)),
+      function(element) {
+        isTRUE(all.equal(x@env[[element]], new_teal_data@env[[element]]))
+      },
+      logical(1)
+    )
+
+    names_diff_other <- setdiff(names(new_teal_data@env), names(x@env))
+
+    error <- c(
+      "Code verification failed at object(s):",
+      paste0("  \u2022 ", c(names_diff, names(which(!objects_diff))))
+    )
+    if (length(names_diff_other)) {
+      error <- c(
+        error,
+        "Object(s) created with code that do not exist in teal_data:",
+        paste0("  \u2022 ", names_diff_other)
+      )
+    }
+
+    stop(error)
```

Comment:
Since we are getting into details, can we explicitly differentiate three categoreis?
1. objects missing (in env but not obtained from code)
2. objects different (in both but fail `all.equal`)
3. excess objects (not in env but obtained from code)

Reply:
Yeah, totally and I think it's gonna be useful for the end user!

Reply:
with a1c5dbfbd68e0e496f80065466f35808c02991cf we ended up with

```r
a <- 1
b <- a+2
c <- list(x = 2)
d <- 5

tdata4 <- teal_data(
  a = a, b = b, c = c, d = d, 
  code = 
    "
    a <- 1
    b <- a
    c <- list(x = 2)
    e <- 1
    "
)

tdata4

# ✖ unverified teal_data object
# <environment: 0x000002d744cd9058> [L]
# Parent: <environment: package:teal.data>
# Bindings:
# • a: <dbl> [L]
# • b: <dbl> [L]
# • c: <named list> [L]
# • d: <dbl> [L]

verify(tdata4)

# Error in verify(tdata4) : Code verification failed.
# Object(s) recreated with code that have different structure in teal_data:
# • b
# Object(s) not created with code that exist in teal_data:
# • d
# Object(s) created with code that do not exist in teal_data:
# • e
```

Reply:
This is great!